provider/openstack/provider.go

```diff
@@ -1039,14 +1040,62 @@ func (e *Environ) StartInstance(args environs.StartInstanceParams) (*environs.St
 		args.InstanceConfig.MachineId,
 	)
 
+	waitForActiveServerDetails := func(
+		client *nova.Client,
+		id string,
+		timeout time.Duration,
+	) (server *nova.ServerDetail, err error) {
+
+		err = retry.Call(retry.CallArgs{
+			Clock: e.clock,
+			Delay: 10 * time.Second,
+			Func: func() error {
+				server, err = client.GetServer(id)
+				if err == nil {
+					if server.Status == nova.StatusBuild {
+						return errors.New("server has status BUILD, waiting 10 seconds before polling again...")
+					}
+				}
+				return nil
```

Comment:
We should be returning `err` here, otherwise we'll ignore any errors from GetServer. You should use the IsFatalError field of CallArgs to decide whether or not to retry. Something like:

```go
var errStillBuilding = errors.New("server is still being built")
err = retry.Call(retry.CallArgs{
    Clock: e.clock,
    Delay: 10 * time.Second,
    MaxDuration: timeout,
    IsFatalError: func(err error) bool {
        return err != errStillBuilding
    },
    NotifyFunc: func(lastError error, attempt int) {
        args.StatusCallback(status.Provisioning, fmt.Sprintf("%q retry attempt %d", lastError, attempt), nil)
    }
    Func: func() error {
        server, err := client.GetServer(id)
        if err != nil {
            return err
        }
        if server.Status == nova.StatusBuild {
            return errStillBuilding
        }
        return nil
    },
})
```